Attempt to detect invalid DB credentials

gregorydlogan opened the pull request:
Adding an attempt to check if the database user permissions are correct.  This is by no means comprehensive, but will catch a bunch of easily missed situations.  The exception looks like the following, and stops Opencast from booting:

```
2020-11-27T16:28:35,989 | ERROR | (BootFeaturesInstaller:113) - Error installing boot features
org.apache.karaf.features.internal.util.MultiException: Error restarting bundles:
        Activator start error in bundle opencast-db [80].
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:1044) ~[?:?]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1062) ~[?:?]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:998) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
        Suppressed: org.osgi.framework.BundleException: Activator start error in bundle opencast-db [80].
                at org.apache.felix.framework.Felix.activateBundle(Felix.java:2290) ~[?:?]
                at org.apache.felix.framework.Felix.startBundle(Felix.java:2146) ~[?:?]
                at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998) ~[?:?]
                at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984) ~[?:?]
                at org.apache.karaf.features.internal.service.BundleInstallSupportImpl.startBundle(BundleInstallSupportImpl.java:165) ~[?:?]
                at org.apache.karaf.features.internal.service.FeaturesServiceImpl.startBundle(FeaturesServiceImpl.java:1153) ~[?:?]
                at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:1036) ~[?:?]
                at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1062) ~[?:?]
                at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:998) ~[?:?]
                at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
                at java.lang.Thread.run(Thread.java:834) [?:?]
        Caused by: java.lang.RuntimeException: Unable to verify SQL credentials have required permissions!
                at org.opencastproject.db.Activator.start(Activator.java:161) ~[?:?]
                at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697) ~[?:?]
                at org.apache.felix.framework.Felix.activateBundle(Felix.java:2240) ~[?:?]
                ... 12 more
```

Note: I've tested against MariaDB 10 so this will probably work against MySQL as well. We should double check that it works against PostgreSQL though... 

### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] be against the correct branch (features can only go into develop)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

gregorydlogan committed:
Adding an attempt to check if the database user permissions are correct.  This is by no means comprehensive, but will catch a bunch of easily missed situations.

lkiesow commented on modules/db/src/main/java/org/opencastproject/db/Activator.java:
```diff
@@ -137,6 +139,36 @@ else if (pooledDataSource.getMaxIdleTime() == 0) {
+    Statement statement = pooledDataSource.getConnection().createStatement();
+
+    long time = System.currentTimeMillis();
+    String tableName = "oc-temp" + time;
```
Nitpicking ;-P
To be consistent with the other table names:

```suggestion
    String tableName = "oc_temp_" + time;
```

lkiesow commented on modules/db/src/main/java/org/opencastproject/db/Activator.java:
```diff
@@ -137,6 +139,36 @@ else if (pooledDataSource.getMaxIdleTime() == 0) {
+    String tableName = "oc-temp" + time;
+    try {
+      runUpdate(statement, "CREATE TABLE " + tableName + " ( id BIGINT NOT NULL, test BIGINT, PRIMARY KEY (id) );");
+      runUpdate(statement, "INSERT INTO " + tableName + " VALUES (" + time + "," + time + ");");
+      runUpdate(statement, "INSERT INTO " + tableName + " VALUES (" + time + 1 + "," + time + 1 + ");");
+      runUpdate(statement, "UPDATE " + tableName + " SET test = " + time + 2 + " WHERE id = " + time + ";");
+      ResultSet rs = statement.executeQuery("SELECT FROM " + tableName + " WHERE id = " + time + 2 + ";");
+      while (rs.next()) {
+        long id = rs.getLong("id");
+        long test = rs.getLong("test");
+        if (id != time || test != time + 2) {
```
I don't think this is actually doing what you want it to do. You are selecting the entry with `id = time + 2` but you only ever set `id = time` and `id = time + 1`. This means that you never actually go into the loop. Also, one simple insert should suffice for this test:

```suggestion
      runUpdate(statement, "INSERT INTO " + tableName + " VALUES (" + time + ", 0);");
      runUpdate(statement, "UPDATE " + tableName + " SET test = " + time + ";");
      ResultSet rs = statement.executeQuery("SELECT FROM " + tableName + ";");
      while (rs.next()) {
        long id = rs.getLong("id");
        long test = rs.getLong("test");
        if (id != time || test != time) {
```

lkiesow commented on modules/db/src/main/java/org/opencastproject/db/Activator.java:
```diff
@@ -137,6 +139,36 @@ else if (pooledDataSource.getMaxIdleTime() == 0) {
       pooledDataSource.getMaxPoolSize(), pooledDataSource.getMinPoolSize(), pooledDataSource.getMaxIdleTime());
+    Statement statement = pooledDataSource.getConnection().createStatement();
+
+    long time = System.currentTimeMillis();
```
I think using a unique table name is a good solution for this problem. But imaging a cluster to start up at the same time. That would potentially select the same table names. Maybe just pick a large random number instead?

lkiesow commented on modules/db/src/main/java/org/opencastproject/db/Activator.java:
```diff
@@ -137,6 +139,36 @@ else if (pooledDataSource.getMaxIdleTime() == 0) {
+      runUpdate(statement, "DELETE FROM " + tableName + " WHERE id = " + time + ";");
+      runUpdate(statement, "DROP TABLE " + tableName + ";");
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to verify SQL credentials have required permissions!");
```
In case of an error, we might leave a random table in the database. We could try deleting it in case of an error?
Something like this in a final block (catching and ignoring excepzions):

```java
runUpdate(statement, "DROP TABLE " + tableName + ";");
```

gregorydlogan committed:
Addressing review comments

gregorydlogan commented:
@lkiesow Good catches.  Patched.

lkiesow reviewed (changes requested):
Unfortunately, running this with the internal h2 database causes:

```
org.apache.karaf.features.internal.util.MultiException: Error restarting bundles:
        Activator start error in bundle opencast-db [80].
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:1044)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1062)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:998)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
        Suppressed: org.osgi.framework.BundleException: Activator start error in bundle opencast-db [80].
                at org.apache.felix.framework.Felix.activateBundle(Felix.java:2290)
                at org.apache.felix.framework.Felix.startBundle(Felix.java:2146)
                at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998)
                at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984)
                at org.apache.karaf.features.internal.service.BundleInstallSupportImpl.startBundle(BundleInstallSupportImpl.java:165)
                at org.apache.karaf.features.internal.service.FeaturesServiceImpl.startBundle(FeaturesServiceImpl.java:1153)
                at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:1036)
                ... 6 more
        Caused by: java.lang.RuntimeException: Unable to verify SQL credentials have required permissions!
                at org.opencastproject.db.Activator.start(Activator.java:160)
                at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697)
                at org.apache.felix.framework.Felix.activateBundle(Felix.java:2240)
                ... 12 more
2020-12-20 23:26:26,321 | ERROR | (BootFeaturesInstaller:113) - Error installing boot features
org.apache.karaf.features.internal.util.MultiException: Error restarting bundles:
        Activator start error in bundle opencast-db [80].
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:1044) ~[?:?]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1062) ~[?:?]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:998) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
        Suppressed: org.osgi.framework.BundleException: Activator start error in bundle opencast-db [80].
                at org.apache.felix.framework.Felix.activateBundle(Felix.java:2290) ~[?:?]
                at org.apache.felix.framework.Felix.startBundle(Felix.java:2146) ~[?:?]
                at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998) ~[?:?]
                at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984) ~[?:?]
                at org.apache.karaf.features.internal.service.BundleInstallSupportImpl.startBundle(BundleInstallSupportImpl.java:165) ~[?:?]
                at org.apache.karaf.features.internal.service.FeaturesServiceImpl.startBundle(FeaturesServiceImpl.java:1153) ~[?:?]
                at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:1036) ~[?:?]
                at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1062) ~[?:?]
                at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:998) ~[?:?]
                at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
                at java.lang.Thread.run(Thread.java:834) [?:?]
        Caused by: java.lang.RuntimeException: Unable to verify SQL credentials have required permissions!
                at org.opencastproject.db.Activator.start(Activator.java:160) ~[?:?]
                at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697) ~[?:?]
                at org.apache.felix.framework.Felix.activateBundle(Felix.java:2240) ~[?:?]
                ... 12 more
```

The stack-trace isn't very helpful. If I should guess the reason, then I would say that you are mixing floating point numbers (`double random`) and integers (database fields are `bigint`) or h2 doesn't like dots in a table.


EDIT: Same error with PostegreSQL.
EDIT2: Same error with MariaDB as well

gregorydlogan committed:
Fixing a bunch of broken logic.

1) double -> long: Table names can't have dots in them (duh).
2) SELECT needs to know what it's selecting.
3) CREATE and DROP table does not generate an affected, yielding a failed test.

gregorydlogan commented:
@lkiesow should be fixed.  A number of issues resolved.  Remember to add CREATE (and DROP) perms to the DB user.

lkiesow reviewed (approved):
Looks good. Will merge this once it's through the round of discussions at the technical meeting.
